File: ruby/lib/coding_interview/application/service/asset_service.rb
require "bigdecimal"

module CodingInterview
  module Application
    module Service
      module AssetService
        module_function

        def evaluate_stock(stock, prices)
          price = prices[stock.symbol]
          raise "missing price for #{stock.symbol}" if price.nil?
          stock.qty * price
        end

        def total_valuation(account, prices)
          account.stocks.inject(account.cash) { |acc, e| acc + evaluate_stock(e, prices) }
        end
      end
    end
  end
end

File: python/src/coding_interview/application/service/asset_service.py
from __future__ import annotations

from decimal import Decimal

from coding_interview.domain.stock import Account, Stock
from coding_interview.domain.stock_symbol import StockSymbol


def evaluate_stock(stock: Stock, prices: dict[StockSymbol, Decimal]) -> Decimal:
    price = prices.get(stock.symbol)
    if price is None:
        raise IllegalStateError(f"missing price for {stock.symbol}")
    return stock.qty * price


def total_valuation(account: Account, prices: dict[StockSymbol, Decimal]) -> Decimal:
    return sum(evaluate_stock(s, prices) for s in account.stocks) + account.cash


class IllegalStateError(Exception):
    pass
